python/pywatchman/pybser.py

```diff
@@ -213,45 +213,106 @@ def _bunser_string(buf, pos):
     return (str_val, pos + str_len)
 
 
-def _bunser_array(buf, pos):
+def _bunser_array(buf, pos, mutable=True):
     arr_len, pos = _bunser_int(buf, pos + 1)
     arr = []
     for i in range(arr_len):
-        arr_item, pos = _bser_loads_recursive(buf, pos)
+        arr_item, pos = _bser_loads_recursive(buf, pos, mutable)
         arr.append(arr_item)
+
+    if not mutable:
+      arr = tuple(arr)
+
     return arr, pos
 
 
-def _bunser_object(buf, pos):
+# This is a quack-alike with the bserObjectType in bser.c
+# It provides by getattr accessors and getitem for both index
+# and name.
+class _BunserDict(object):
```

Comment:
How about `collections.Mapping` instead? It's immutable and will provide you with more useful dict-like methods.